Refactoring source organization, tally initialization, global variables

paulromano opened the pull request:
This branch heavily refactors source organization, mostly related to global variables, tallies, and initialization. Given how large the diff is, it might be good to first discuss why I'm making all these changes. There are three motivations:

1. I'm working on incorporating Xingjie Peng's work on calculating sensitivities in OpenMC. A lot of what he did worked very independently of the existing tally system, which was good in one sense but it also resulted in a lot of duplicated code. A lot of the tally refactoring as part of this PR is aimed at making `TallyObject` more suitable as a "parent class" of a derived tally.

2. I'm also working on changing how [OpenDeplete](https://github.com/mit-crpg/opendeplete) does coupling from filesystem-based to an in-memory coupling. The first step of this was to extend the C API and corresponding Python bindings so that tallies can be created in memory from Python without ever writing/reading an XML file. This PR achieves that. Over time, I intend to have our C API capable enough to avoid XML altogether; one step at a time though.

3. With an eye to the future, a lot of the changes here will, I hope, ease an eventual transition to C++.

To highlight the changes:

- All module variables in global.F90 have been split up over many modules. This was done primarily to eliminate circular dependencies. I've tried to put arrays of a given type in the module where the type is actually defined (e.g., `type(Cell) :: cells(:)` in geometry_header.F90).
- A new `simulation_header` module includes global variables related to things that are used/are changing during a simulation as opposed.
- A new `settings` module includes global variables that store settings read from settings.xml. Separating settings from simulation-related variables also helps to avoid circular dependencies.
- Some redundant global variables have been eliminated, e.g., `active_batches` and `tallies_on` don't give any extra information that can be inferred from other global variables, so they were removed.
- I've moved all tally-related modules into a tallies/ subdirectory of src/. I'm planning on doing this with other areas too to make it clearer where to find specific functionality.
- I've tried to move a lot of XML-reading functionality into type-bound procedures for specific derived types rather than all being defined in input_xml. My eventual goal is to have input_xml be very lean, just calling from_xml methods for the most part. The secondary effect of this is that it will make it much easier to translate a lot of the source to another language (C++) since the derived types on the Fortran side will contain all the relevant code rather than being spread out across many modules.
- Tally initialization has been substantially refactored. Since tallies can be added at various points (e.g., CMFD adds its own tallies), I've added a bunch of C API calls that allow this to be done in a controlled many. For example, if you want to allocate two new tallies, you'd call `openmc_extend_tallies`.
- Tally filters have been separated into their own modules, and importantly they don't depend on the `global` module anymore.
- I've changed how entropy and UFS meshes are specified. Now, you can give a `<mesh>` element in settings.xml and a new `<entropy_mesh>` element just specifies the ID of a given mesh. Same goes for UFS. This allows the same meshes to be used for things in settings as well as in tallies. It also reduces the amount of code needed to handle meshes since all the entropy- and UFS-specific mesh reading code is now gone.
- Fixed an issue with the Python bindings to the C API not getting installed properly (thanks @cjosey for finding that).
- A lot of the procedures that were in the `initialize` module have been moved to `input_xml` because they are really called when XML files are being read. The `initialize` module now contains only procedures that initialize the state of the program (rather than the state of a user's model)
- Refactored creation of CMFD tallies to use all C API function calls.

bhermanmit commented on src/cmfd_execute.F90:
```diff
@@ -276,10 +271,21 @@ subroutine cmfd_reweight(new_weights)
+              if (cmfd % cmfd_src(g,ijk(1),ijk(2),ijk(3)) > ZERO) then
+                cmfd % weightfactors(g,ijk(1),ijk(2),ijk(3)) = &
+                     cmfd % cmfd_src(g,ijk(1),ijk(2),ijk(3)) * norm &
+                     / cmfd % sourcecounts(ng - g + 1, mesh_bin)
```
@paulromano - what was the motivation for changing this, just curious?

paulromano replied:
I ended up changing this when I was changing how `count_bank_sites` worked. Before, as you'll recall, it operated on a 4D array where the dimension corresponded to an energy grid and the x- y- and z- dimensions of a mesh. I wanted it to work instead on a 2D array of (energy, mesh bin) so that it potentially could work with any type of mesh. That caused `cmfd % sourcecounts` to become a 2D array instead of 4D. As a result, it wasn't possible to use the `WHERE` block to calculate `cmfd % cmfd_src` because it was still 4D whereas `sourcecounts` was 2D, so it got changed to nested do loops.

smharper commented on src/tallies/tally_filter_header.F90:
```diff
@@ -104,4 +129,95 @@ subroutine filter_initialize(this)
+    ! Return indices in filters array
+    if (present(index_start)) index_start = n_filters + 1
+    if (present(index_end)) index_end = n_filters + n
+    n_filters = n_filters + n
```
Perhaps we should remove the `n_filters` variable and just use `size(filters)` to avoid inconsistent data.  Or maybe make an `n_filters()` function to handle an `if (allocated)`

paulromano replied:
I've kind of gone back and forth on this. At one point I wanted to get rid of all the `n_` variables. However, the one nice thing about it is that you can use `n_filters` when there are zero filters, whereas `size(filters)` will cause trouble if `filters` hasn't been allocated. So, now I lean towards leaving this as is because it provides an easy way to check the size of an array from Python/ctypes.

smharper replied:
Good point.  What do you think of allocating arrays like this with length 0?  I think this might already be the case with tally derivatives.

paulromano replied:
Yes, that would be another option. The only problem I see with that is that there is no way to have arrays allocated to size 0 at program startup. That could be done on the call to `openmc_init`, though. In terms of complexity, having `n_*` be functions that look at `size(...)` would probably result in more lines than the current approach. With a good API, there should only be a few spots where allocations happen (and hence where a `n_*` global variable would change).

smharper commented on src/tallies/tally_filter_header.F90:
```diff
@@ -104,4 +129,95 @@ subroutine filter_initialize(this)
+  end function openmc_filter_set_id
+
+
+  function openmc_get_filter(id, index) result(err) bind(C)
```
`openmc_index_filter` would be a more specific name.

Similarly, `openmc_index_tally`, `openmc_index_cell`, ...

paulromano replied:
I've renamed these to `openmc_get_*_index`.

paulromano commented on openmc/stats/multivariate.py:
```diff
@@ -12,6 +12,10 @@
 
 
+__all__ = ['UnitSphere', 'PolarAzimuthal', 'Isotropic', 'Monodirectional',
+           'Spatial', 'CartesianIndependent', 'Box', 'Point']
```
`__all__` specifies what attributes are imported whenever you have a `from _ import *` statement, so in general the purpose is to expose only certain attributes. My intent was really to avoid polluting namespaces, but since this only goes like 1% of the way (since it's not being used elsewhere), I'll just hold off and remove this. We can debate the merits of `__all__` in a separate PR.

paulromano commented:
A note about my latest commits -- I've reworked error handling a bit because I was getting worried about the number of error codes exploding as the C API grows larger. I've changed it so that there are a few general categories of error codes as well as a global C interoperable error message that can be accessed from Python/ctypes. So, now when a function from the C API returns an error (negative integer), Python raises an exception corresponding to the error code and uses the error message that was set on the Fortran side.

paulromano commented:
One more change I made -- I've removed the 'iso-in-lab' option for nuclides in a material because it requires storing a logical flag on every nuclide in every material. In Fortran, logicals are essentially 4-byte integers so the memory requirement may end up being substantial for large models with depleted materials.

@smharper I believe all your comments have been addressed. Let me know if you have further requests for changes.

smharper commented:
iso-in-lab is a really useful feature, though.  Can you leave it in, but put preprocessor `#ifdef`s around the relevant code?

smharper commented:
Looks good!  Merged
